app/src/main/java/uk/gov/onelogin/login/state/IStateGenerator.kt

```diff
@@ -0,0 +1,5 @@
+package uk.gov.onelogin.login.state
+
+interface IStateGenerator {
+    fun generate(): String
+}
```

Comment:
As this only has one function, we're able to convert this into a functional interface. This has the real nice benefit of how you can go about creating anonymous implementations - `val something = IStateGenerator { "returnSomeString" }`.

We can then declare test implementations like so:

```suggestion
fun interface IStateGenerator {
    fun generate(): String
    
    companion object {
    
        /**
         * Create an [IStateGenerator] instance that returns the provided [state]
         * for the [IStateGenerator.generate] function.
         *
         * Use within tests for providing custom data.
         *
         * @return The [state] parameter.
         */
        @VisibleForTesting(otherwise = VisibleForTesting.private)
        fun stub(state: String) = IStateGenerator { state }
        
        /**
         * Create an [IStateGenerator] instance that returns the [state.invoke] value
         * for the [IStateGenerator.generate] function.
         *
         * Use within tests for providing custom data.
         *
         * @return The [state.invoke] return value.
         */
        @VisibleForTesting(otherwise = VisibleForTesting.private)
        fun stub(state: () -> String) = IStateGenerator { state.invoke() }
        
        /**
         * Create an [IStateGenerator] instance that causes an [Exception] upon use.
         *
         * Use within tests that don't require [IStateGenerator] usage whilst still needing
         * the presence of a generator as a way of marking those tests as unrelated.
         */
        @VisibleForTesting(otherwise = VisibleForTesting.private)
        fun dummy() = IStateGenerator {
            throw IllegalAccessException(
                "This IStateGenerator instance should not have been used!"
            )
        }
    }
}
```

Do you think there's value in having the preceding companion object? If so, do we want to do this with other interfaces that have a single function within them?  I like that there's only one function in things

Reply:
In jira ticket